File: Objective-C/Reference App/401. Report Driver Current Location/ReportDriverCurrentLocationViewController.m
//
//  ReportDriverCurrentLocationViewController.m
//  Reference App
//
//  Created by David Deller on 5/3/18.
//  Copyright © 2018 TallyGo. All rights reserved.
//

#import "ReportDriverCurrentLocationViewController.h"

@interface ReportDriverCurrentLocationViewController ()

@property (nonatomic, weak) IBOutlet UILabel *serverURLLabel;

@property (nonatomic, strong, nullable) TGTurnByTurnViewController *turnByTurnViewController;

// Put these inside your @interface
// Keep track of when and what we last reported.
@property (nonatomic, nullable) TGPoint *lastReportedLocationPoint;
@property (nonatomic, nullable) NSDate *lastReportedLocationTime;

@end

@implementation ReportDriverCurrentLocationViewController

// Put these inside your @implementation
// You can change these configuration values to suit your needs.
static CLLocationDistance const reportLocationPointTrigger = 100;
static NSTimeInterval const reportLocationTimeTrigger = 60;
static NSString *const reportLocationURL = @"http://localhost:3200/drivers/current_location";
static NSString *const reportLocationMethod = @"PUT";

- (void)viewDidLoad {
    [super viewDidLoad];
    
    NSURL *URL = [NSURL URLWithString:reportLocationURL];
    self.serverURLLabel.text = [NSString stringWithFormat:@"%@://%@:%@", URL.scheme, URL.host, URL.port];
}

- (IBAction)go:(id)sender {
    [self startTurnByTurn];
}

- (void)startTurnByTurn {
    TallyGo.simulatedCoordinate = CLLocationCoordinate2DMake(34.101558, -118.340944); // Grauman's Chinese Theatre
    
    // Get these coordinates from your app, these are just a sample
    TGWaypoint *origin = [TGWaypoint.alloc initWithCoordinate:CLLocationCoordinate2DMake(34.101558, -118.340944)]; // Grauman's Chinese Theatre
    TGWaypoint *destination = [TGWaypoint.alloc initWithCoordinate:CLLocationCoordinate2DMake(34.011441, -118.494932)]; // Santa Monica Pier
    
    // Configure turn-by-turn navigation
    TGTurnByTurnConfiguration *config = [TGTurnByTurnConfiguration new];
    config.showsOriginIcon = NO;
    config.originWaypoint = origin;
    config.destinationWaypoint = destination;
    config.commencementSpeech = @"Let's go!";
    config.proceedToRouteSpeech = @"Please proceed to the route.";
    config.arrivalSpeech = @"You have arrived.";
    
    // If you’d rather skip the preview and jump straight into directions, you can do that:
    TGTurnByTurnViewController *viewController = [TGTurnByTurnViewController createWithConfiguration:config];
    [self presentViewController:viewController animated:YES completion:nil];
    
    self.turnByTurnViewController = viewController;
    [self setupReporting];
}

- (void)setupReporting {
    [NSNotificationCenter.defaultCenter addObserverForName:TGTelemetryCurrentPointChangeNotification object:nil queue:nil usingBlock:^(NSNotification * _Nonnull notification) {
        TGPoint *newPoint = notification.userInfo[TGTelemetryUserInfoToPoint];
        if (newPoint == nil) {
            return;
        }
        
        [self reportLocation:newPoint];
    }];
}

- (void)reportLocation:(nonnull TGPoint *)point {
    // We should only continue if we haven't reported yet, or if one of our triggers has been satisfied.
    if (self.lastReportedLocationPoint != nil &&
        self.lastReportedLocationTime != nil &&
        TGLocationCoordinateDistanceFromCoordinate(self.lastReportedLocationPoint.coordinate, point.coordinate) <= reportLocationPointTrigger &&
        [NSDate.date timeIntervalSinceDate:self.lastReportedLocationTime] <= reportLocationTimeTrigger) {
        
        return;
    }
    
    // Collect the data we want to send
    NSDictionary *requestData = @{
        @"latitude": @(point.coordinate.latitude),
        @"longitude": @(point.coordinate.longitude),
    };
    
    // Configure the request
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:reportLocationURL]];
    request.HTTPMethod = reportLocationMethod;
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    
    NSError *error = nil;
    NSData *encodedData = [NSJSONSerialization dataWithJSONObject:requestData options:0 error:&error];
    if (encodedData != nil) {
        request.HTTPBody = encodedData;
        
    } else {
        @throw error;
    }
    
    // Send the request
    NSURLSessionDataTask *task = [NSURLSession.sharedSession dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        
        if (error != nil) {
            NSLog(@"Error reporting current location: %@", error.userInfo);
            [self handleError:error viewController:self.turnByTurnViewController];
            
        } else {
            // Optionally, you can do something with the successful response here.
        }
    }];
    [task resume];
    
    // Remember that we did this
    self.lastReportedLocationPoint = point;
    self.lastReportedLocationTime = NSDate.date;
}

@end

File: Swift 4/Reference App/401. Report Driver Current Location/ReportDriverCurrentLocationViewController.swift
//
//  ReportDriverCurrentLocationViewController.swift
//  Reference App
//
//  Created by David (work 2) on 5/2/18.
//  Copyright © 2018 TallyGo. All rights reserved.
//

import UIKit
import TallyGoKit

class ReportDriverCurrentLocationViewController: ExampleViewController {

    @IBOutlet weak var serverURLLabel: UILabel!
    
    var turnByTurnViewController: TGTurnByTurnViewController?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        serverURLLabel.text = "\(reportLocationURL.scheme!)://\(reportLocationURL.host!):\(reportLocationURL.port!)"
    }
    
    @IBAction func go(_ sender: Any) {
        startTurnByTurn()
    }
    
    func startTurnByTurn() {
        TallyGo.simulatedCoordinate = CLLocationCoordinate2D(latitude: 34.101558, longitude: -118.340944) // Grauman's Chinese Theatre
        
        // Get these coordinates from your app, these are just a sample
        let origin = CLLocationCoordinate2D(latitude: 34.101558, longitude: -118.340944) // Grauman's Chinese Theatre
        let destination = CLLocationCoordinate2D(latitude: 34.011441, longitude: -118.494932) // Santa Monica Pier
        
        // Configure turn-by-turn navigation
        let config = TGTurnByTurnConfiguration()
        config.showsOriginIcon = false
        config.origin = origin
        config.destination = destination
        config.commencementSpeech = "Let's go!"
        config.proceedToRouteSpeech = "Please proceed to the route."
        config.arrivalSpeech = "You have arrived."
        
        // If you’d rather skip the preview and jump straight into directions, you can do that:
        let viewController = TGTurnByTurnViewController.create(with: config)
        present(viewController, animated: true, completion: nil)
        
        turnByTurnViewController = viewController
        setupReporting()
    }
    
    func setupReporting() {
        NotificationCenter.default.addObserver(forName: NSNotification.Name.TGTelemetryCurrentPointChange, object: nil, queue: nil) { (notification) in
            guard let newPoint = notification.userInfo?[TGTelemetryUserInfo.toPoint] as? TGPoint else {
                return
            }
            
            self.reportLocation(newPoint)
        }
    }
    
    // You can change these configuration values to suit your needs.
    let reportLocationPointTrigger: CLLocationDistance = 100
    let reportLocationTimeTrigger: TimeInterval = 60
    let reportLocationURL = URL(string: "http://localhost:3200/drivers/current_location")!
    let reportLocationMethod = "PUT"
    
    // Keep track of when and what we last reported.
    var lastReportedLocationPoint: TGPoint?
    var lastReportedLocationTime: Date?
    
    func reportLocation(_ point: TGPoint) {
        // We should only continue if we haven't reported yet, or if one of our triggers has been satisfied.
        guard (lastReportedLocationPoint == nil ||
            lastReportedLocationTime == nil ||
            TGLocationCoordinateDistanceFromCoordinate(lastReportedLocationPoint!.coordinate, point.coordinate) > reportLocationPointTrigger ||
            Date().timeIntervalSince(lastReportedLocationTime!) > reportLocationTimeTrigger) else {
                
            return
        }
        
        // Collect the data we want to send
        let requestData = [
            "latitude": point.coordinate.latitude,
            "longitude": point.coordinate.longitude,
        ]
        
        // Configure the request
        var request = URLRequest(url: reportLocationURL)
        request.httpMethod = reportLocationMethod
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try! JSONEncoder().encode(requestData)
        
        // Send the request
        let task = URLSession.shared.dataTask(with: request) { (data, response, error) in
            if let error = error {
                NSLog("Error reporting current location: \(error)")
                self.handleError(error, viewController: self.turnByTurnViewController)
                
            } else {
                // Optionally, you can do something with the successful response here.
            }
        }
        task.resume()
        
        // Remember that we did this
        lastReportedLocationPoint = point
        lastReportedLocationTime = Date()
    }
    
}
